drivers/sdhc/sdhc_esp32.c

```diff
@@ -0,0 +1,1544 @@
+		ios->clock, ios->power_mode == SDHC_POWER_ON ? "ON" : "OFF",
+		ios->signal_voltage == SD_VOL_1_8_V ? "1.8V" : "3.3V");
+
+	if (ios->clock) {
+		/* Check for frequency boundaries supported by host */
+		if (ios->clock > data->props.f_max || ios->clock < data->props.f_min) {
+			LOG_ERR("Proposed clock outside supported host range");
+			return -EINVAL;
+		}
+
+		if (data->bus_clock != (uint32_t)ios->clock) {
+			/* Try setting new clock */
+			ret = sdmmc_host_set_card_clk(cfg->slot, (ios->clock / 1000));
+
+			if (ret == 0) {
+				LOG_INF("Bus clock successfully set to %d kHz", ios->clock / 1000);
+			} else {
+				LOG_ERR("Error configuring card clock");
+				return err_esp2zep(ret);
+			}
+
+			data->bus_clock = (uint32_t)ios->clock;
+		}
+	}
+
+	if (ios->bus_width > 0) {
+		/* Set bus width */
+		switch (ios->bus_width) {
+		case SDHC_BUS_WIDTH1BIT:
+			bus_width = 1;
+			break;
+		case SDHC_BUS_WIDTH4BIT:
+			bus_width = 4;
+			break;
+		default:
+			return -ENOTSUP;
+		}
+
+		if (data->bus_width != bus_width) {
+			ret = sdmmc_host_set_bus_width(cfg->slot, bus_width);
+
+			if (ret == 0) {
+				LOG_INF("Bus width set successfully to %d bit", bus_width);
+			} else {
+				LOG_ERR("Error configuring bus width");
+				return err_esp2zep(ret);
+			}
+
+			data->bus_width = bus_width;
+		}
+	}
+
+	/* Toggle card power supply */
+	if ((data->power_mode != ios->power_mode) && (cfg->pwr_gpio.port)) {
+		if (ios->power_mode == SDHC_POWER_OFF) {
+			gpio_pin_set_dt(&cfg->pwr_gpio, 0);
+		} else if (ios->power_mode == SDHC_POWER_ON) {
+			gpio_pin_set_dt(&cfg->pwr_gpio, 1);
+		}
+		data->power_mode = ios->power_mode;
+	}
+
+	if (ios->timing > 0) {
+		/* Set I/O timing */
+		if (data->timing != ios->timing) {
+			switch (ios->timing) {
+			case SDHC_TIMING_LEGACY:
+			case SDHC_TIMING_HS:
+				sdmmc_ll_enable_ddr_mode(sdio_hw, cfg->slot, false);
+				break;
+			case SDHC_TIMING_DDR50:
+			case SDHC_TIMING_DDR52:
+				/* Enable DDR mode */
+				sdmmc_ll_enable_ddr_mode(sdio_hw, cfg->slot, true);
+				LOG_INF("DDR mode enabled");
+				break;
+			case SDHC_TIMING_SDR12:
+			case SDHC_TIMING_SDR25:
+				sdmmc_ll_enable_ddr_mode(sdio_hw, cfg->slot, false);
+				break;
+			case SDHC_TIMING_SDR50:
+			case SDHC_TIMING_HS400:
+			case SDHC_TIMING_SDR104:
+			case SDHC_TIMING_HS200:
+			default:
+				LOG_ERR("Timing mode not supported for this device");
+				ret = -ENOTSUP;
+				break;
+			}
+
+			LOG_INF("Bus timing successfully changed to %s", timingStr[ios->timing]);
+			data->timing = ios->timing;
+		}
+	}
+
+	return ret;
+}
+
+/*
+ * Return 0 if card is not busy, 1 if it is
+ */
+static int sdhc_esp32_card_busy(const struct device *dev)
+{
+	const struct sdhc_esp32_config *cfg = dev->config;
+	const sdmmc_dev_t *sdio_hw = cfg->sdio_hw;
+
+	return (sdio_hw->status.data_busy == 1);
+}
+
+/*
+ * Execute card tuning
+ */
+static int sdhc_esp32_execute_tuning(const struct device *dev)
```

Comment:
You can omit the implementation of this function entirely, the API wrapper for `sdhc_execute_tuning` will report `-ENOSYS` if a function pointer for tuning is not present in the API structure